neqo-transport/src/connection/tests/recovery.rs

```diff
@@ -944,3 +944,99 @@ fn ack_for_unsent() {
         }
     ));
 }
+
+/// Test that PTO fires for Handshake space even when no Handshake packets have been sent.
+///
+/// This reproduces the haproxy handshakeloss bug where:
+/// 1. Client receives Server Hello → Handshake keys installed
+/// 2. Server's Handshake flight (Certificate, etc.) is lost/corrupted
+/// 3. Client never sends any Handshake packets (has nothing to send yet)
+/// 4. Client's Handshake space has last_ack_eliciting = None
+/// 5. PTO timer never arms for Handshake space
+/// 6. Connection times out
+///
+/// RFC 9002 Section 6.2.4: "When a PTO timer expires, a sender MUST send at least
+/// one ack-eliciting packet in the packet number space as a probe."
+///
+/// The client should probe Handshake space to elicit server retransmission.
+#[test]
+fn pto_handshake_space_when_server_flight_lost() {
+    const RTT: Duration = Duration::from_millis(10);
+    let mut now = now();
+    // This test assumes PTOs only involve single-packet flights, which is incompatible with
+    // multi-packet flights as used by MLKEM.
+    let mut client = new_client(ConnectionParameters::default().mlkem(false));
+    let mut server = default_server();
+    // This is a greasing transport parameter, and large enough that the
+    // server needs to send two Handshake packets.
+    let big = TransportParameter::Bytes(vec![0; Pmtud::default_plpmtu(DEFAULT_ADDR.ip())]);
+    server
+        .set_local_tparam(TestTransportParameter, big)
+        .unwrap();
+
+    let c1 = client.process_output(now).dgram();
+
+    now += RTT / 2;
+    let s1 = server.process(c1, now).dgram();
+    assert!(s1.is_some());
+    let s2 = server.process_output(now).dgram();
+    assert!(s2.is_some());
+
+    // Now let the client have the Initial, but drop the first coalesced Handshake packet.
+    now += RTT / 2;
+    let (initial, _) = split_datagram(&s1.unwrap());
+    client.process_input(initial, now);
```

Comment:
So this is going to cause problems when (not if) the server Initial gets too big to fit into a single datagram.

The problem here is that we are on the boundary here and while the first datagram contains Handshake packets NOW, that is not going to remain the case for long.

The complicated part is that you need to take the second packet, TEST whether it is an initial packet, and then split it conditionally.